**Load Basic Libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

In [ ]:
from google.colab import files
f=files.upload()

Saving heart.csv to heart.csv


In [ ]:
df=pd.read_csv( 'heart.csv')
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      52    1   0       125   212    0        1      168      0      1.0   
1      53    1   0       140   203    1        0      155      1      3.1   
2      70    1   0       145   174    0        1      125      1      2.6   
3      61    1   0       148   203    0        1      161      0      0.0   
4      62    0   0       138   294    1        1      106      0      1.9   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  target  
0         2   2     3       0  
1         0   0     3       0  
2         0   0     3       0  
3         2   1     3       0  
4         1   3     2       0  
...     ...  ..   ...     ...  
1020      2   0     2       1  
1021      1   1     3       0  
1022      1   1     2       0  
1023      2   0     2       1  
1024      1   1     3       0  

[1025 rows x 14 columns]

In [ ]:
df.head()#first 5 rows of  a data set

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [ ]:
df.tail()#last 5 rows of  a data set

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  target  
1020      2   0     2       1  
1021      1   1     3       0  
1022      1   1     2       0  
1023      2   0     2       1  
1024      1   1     3       0

**Exploring the data set in order to derive useful information**

In [ ]:
df.shape #how many  rows and  columns

(1025, 14)

In [ ]:
df.columns #we can see the column names here

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [ ]:
"""
Statstical Details Describe provides us with statistical information in the numerical format.
we can infer that in the AGE column the minimum age is 29yrs and maximium is 77yrs mean of age is 54yrs.
The quartiles details are given in form of 25%, 50% and 75%.
The data is divided into 3 quartiles or 4 equal parts. so 25% values lie in each group.
standard deviation and mean are statistical measures which give us an idea of the central tendency of the data set.
However , mean is effected by outliers and hence we need more information to make accurate decisions.
"""
df.describe()#statistical measures of the data

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

In [ ]:
df.isnull().sum()#checking for missing values

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [ ]:
"""
we notice that the dataset has no null values.
this saved us from conveting the null values into some data or dropping then altogether.
"""

print(df.info())#information of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB
None


**Finding the correlation among the attributes**

In [ ]:
df['sex'].value_counts() #1--male,0-->female

1    713
0    312
Name: sex, dtype: int64

In [ ]:
df['target'].value_counts()

1    526
0    499
Name: target, dtype: int64

**fbs(fasting blood sugar > 120 mg/dl) (1 = true; 0 = false) and chest pain relation**

In [ ]:
df['fbs'].value_counts()

0    872
1    153
Name: fbs, dtype: int64

**Preparing the data for Model**

Scaling the data

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X= df.drop(['target'], axis=1)
y= df['target']

In [ ]:
X_train, X_test,y_train, y_test=train_test_split(X,y,test_size=0.3,random_state=40)

**Check the sample Size**

In [ ]:
print('X_train-', X_train.size)
print('X_test-',X_test.size)
print('y_train-', y_train.size)
print('y_test-', y_test.size)

X_train- 9321
X_test- 4004
y_train- 717
y_test- 308


**Logistic Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()

model1=lr.fit(X_train,y_train)
prediction1=model1.predict(X_test)
prediction1_train=model1.predict(X_train)

In [ ]:
X_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
1003   52    1   0       108   233    1        1      147      0      0.1   
758    46    0   1       105   204    0        1      172      0      0.0   
498    54    1   0       140   239    0        1      160      0      1.2   
715    70    1   1       156   245    0        0      143      0      0.0   
1008   42    1   1       120   295    0        1      162      0      0.0   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
942    69    0   3       140   239    0        1      151      0      1.8   
212    50    1   0       150   243    0        0      128      0      2.6   
530    60    0   0       150   258    0        0      157      0      2.6   
167    57    0   0       120   354    0        1      163      1      0.6   
630    45    1   3       110   264    0        1      132      0      1.2   

      slope  ca  thal  
1003      2   3     3  
758       2   0     2  
498       2   0     2  
715       2   0     2  
1008      2   0     2  
...     ...  ..   ...  
942       2   2     2  
212       1   0     3  
530       1   2     3  
167       2   0     2  
630       1   0     3  

[308 rows x 13 columns]

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,prediction1)

0.8409090909090909

In [ ]:
accuracy_score(y_train,prediction1_train)

0.8535564853556485

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, prediction1))

              precision    recall  f1-score   support

           0       0.89      0.77      0.83       153
           1       0.80      0.91      0.85       155

    accuracy                           0.84       308
   macro avg       0.85      0.84      0.84       308
weighted avg       0.85      0.84      0.84       308



**Decision Tree**

In [ ]:
from sklearn.tree import DecisionTreeClassifier

dtc=DecisionTreeClassifier()
model2=dtc.fit(X_train,y_train)
prediction2=model2.predict(X_test)
prediction2_train=model2.predict(X_train)

In [ ]:
accuracy_score(y_test,prediction2)

0.9805194805194806

In [ ]:
accuracy_score(y_train,prediction2_train)

1.0

In [ ]:
print(classification_report(y_test, prediction2))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98       153
           1       1.00      0.96      0.98       155

    accuracy                           0.98       308
   macro avg       0.98      0.98      0.98       308
weighted avg       0.98      0.98      0.98       308



**Random Forest**

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rfc=RandomForestClassifier()
model3 = rfc.fit(X_train, y_train)
prediction3 = model3.predict(X_test)
prediction3_train= model3.predict(X_train)

In [ ]:
accuracy_score(y_test, prediction3)

1.0

In [ ]:
accuracy_score(y_train, prediction3_train)

1.0

In [ ]:
print(classification_report(y_test, prediction3))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       153
           1       1.00      1.00      1.00       155

    accuracy                           1.00       308
   macro avg       1.00      1.00      1.00       308
weighted avg       1.00      1.00      1.00       308



In [ ]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

svm:


In [ ]:
from sklearn.svm import SVC

svm=SVC()
model4=svm.fit(X_train,y_train)
prediction4=model4.predict(X_test)
prediction4_train=model4.predict(X_train)

In [ ]:
accuracy_score(y_test, prediction4)

0.6688311688311688

In [ ]:
accuracy_score(y_train, prediction4_train)

0.702928870292887

Naive_bayes:

In [ ]:
from sklearn.naive_bayes import GaussianNB

NB = GaussianNB()
model5 = NB.fit(X_train, y_train)
prediction5 = model5.predict(X_test)
prediction5_train= model5.predict(X_train)

In [ ]:
accuracy_score(y_test, prediction5)

0.8051948051948052

In [ ]:
accuracy_score(y_train, prediction5_train)

0.8312412831241283

KNN:

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

KNN = KNeighborsClassifier()
model6 = KNN.fit(X_train, y_train)
prediction6 = model6.predict(X_test)
prediction6_train = model6.predict(X_train)

In [ ]:
print('KNN :', accuracy_score(y_test, prediction6))
print('lr :', accuracy_score(y_test, prediction1))
print('dtc :', accuracy_score(y_test, prediction2))
print('rfc :', accuracy_score(y_test, prediction3))
print('NB: ', accuracy_score(y_test, prediction4))
print('SVM :', accuracy_score(y_test, prediction5))

KNN : 0.7142857142857143
lr : 0.8409090909090909
dtc : 0.9805194805194806
rfc : 1.0
NB:  0.6688311688311688
SVC : 0.8051948051948052


In [ ]:
print('KNN :', accuracy_score(y_train, prediction6_train))
print('lr :', accuracy_score(y_train, prediction1_train))
print('dtc :', accuracy_score(y_train, prediction2_train))
print('rfc :', accuracy_score(y_train, prediction3_train))
print('NB: ', accuracy_score(y_train, prediction4_train))
print('SVM :', accuracy_score(y_train, prediction5_train))

KNN : 0.8758716875871687
lr : 0.8535564853556485
dtc : 1.0
rfc : 1.0
NB:  0.702928870292887
SVC : 0.8312412831241283


In [ ]:
input_data=[    50,     0,  0,  110,    254,    0,  0,  159,    0 , 0.0 ,   2,  0 , 2       ]

In [ ]:
 input_data_as_numpy_array=np.asarray(input_data)

In [ ]:
input_data_reshaped=input_data_as_numpy_array.reshape(1,-1)

In [ ]:
prediction= model5.predict(input_data_reshaped)

In [ ]:
print(prediction)#1-->heart disease,0-->not having heart disease

[1]


In [ ]:
if(prediction[0]==0):
   print("The person must not have heart disease")
else:
   print("The person has heart disease")

The person has heart disease


**CONCLUSION**



1.   kNN is slightly underfitted
2.   Logistic Regression ,Naive Bayes and SVM are fitted normally
3.decision tree and Random Forest  are overfitted



***LogisticRegression got best accuracy 84%***